File: src/acp.rs
//! ACP (Agent Client Protocol) client — JSON-RPC 2.0 over stdio.
//!
//! Spawns `gemini.cmd --acp` and provides a typed session API.
//!
//! # Protocol flow
//!
//! 1. `initialize` — handshake, get capabilities
//! 2. `session/new` — create session, get sessionId
//! 3. `session/prompt` — send prompt, receive streaming chunks + final result

use std::io::{BufRead, BufReader, Write};
use std::path::Path;
use std::process::{Child, ChildStdin, Command, Stdio};

use serde_json::Value;

/// A JSON-RPC error from the agent.
#[derive(Debug)]
pub struct AcpError(pub String);

impl std::fmt::Display for AcpError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for AcpError {}

// === Pure functions: testable without a process ===

/// Build a JSON-RPC 2.0 request message.
pub fn build_request(id: u64, method: &str, params: Value) -> String {
    let msg = serde_json::json!({
        "jsonrpc": "2.0",
        "id": id,
        "method": method,
        "params": params
    });
    serde_json::to_string(&msg).expect("serialize JSON-RPC")
}

/// Build the initialize request params.
pub fn build_initialize_params() -> Value {
    serde_json::json!({
        "protocolVersion": 1,
        "clientInfo": {"name": "resident-ai", "version": "0.1.0"}
    })
}

/// Build the session/new request params.
pub fn build_session_new_params(cwd: &str) -> Value {
    serde_json::json!({
        "cwd": cwd,
        "mcpServers": []
    })
}

/// Build the session/prompt request params.
pub fn build_prompt_params(session_id: &str, text: &str) -> Value {
    serde_json::json!({
        "sessionId": session_id,
        "prompt": [{"type": "text", "text": text}]
    })
}

/// Build prompt text with @file references prepended.
pub fn build_prompt_text(text: &str, files: Option<&[&str]>) -> String {
    match files {
        Some(files) if !files.is_empty() => {
            let refs: String = files.iter().map(|f| format!("@{} ", f)).collect();
            format!("{}{}", refs, text)
        }
        _ => text.to_string(),
    }
}

/// Extract sessionId from a session/new response.
pub fn extract_session_id(response: &Value) -> Option<&str> {
    response.get("result")
        .and_then(|r| r.get("sessionId"))
        .and_then(|v| v.as_str())
}

/// Extract agent_message_chunk text from a session/update notification.
/// Returns None if the line is not an agent_message_chunk.
pub fn extract_chunk_text(line: &Value) -> Option<&str> {
    let update = line.get("params")?.get("update")?;
    if update.get("sessionUpdate")?.as_str()? != "agent_message_chunk" {
        return None;
    }
    update.get("content")?.get("text")?.as_str()
}

/// Check if a JSON-RPC line is a response with the given id.
pub fn is_response(line: &Value, expected_id: u64) -> bool {
    line.get("id").and_then(|v| v.as_u64()) == Some(expected_id)
}

/// Check if a JSON-RPC response contains an error.
pub fn response_error(line: &Value) -> Option<&Value> {
    line.get("error")
}

// === AcpSession: process-dependent ===

/// An active ACP session with a running gemini process.
pub struct AcpSession {
    child: Child,
    stdin: ChildStdin,
    reader: BufReader<std::process::ChildStdout>,
    session_id: String,
    next_id: u64,
}

impl AcpSession {
    /// Spawn gemini.cmd --acp, perform handshake, return ready session.
    pub fn new(cwd: &Path) -> Result<Self, AcpError> {
        let mut child = Command::new("gemini.cmd")
            .arg("--acp")
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| AcpError(format!("spawn gemini.cmd --acp: {}", e)))?;

        let stdin = child.stdin.take()
            .ok_or_else(|| AcpError("no stdin".into()))?;
        let stdout = child.stdout.take()
            .ok_or_else(|| AcpError("no stdout".into()))?;

        let mut s = AcpSession {
            child,
            stdin,
            reader: BufReader::new(stdout),
            session_id: String::new(),
            next_id: 1,
        };

        // initialize
        let id = s.send("initialize", build_initialize_params())?;
        let resp = s.read_until_id(id)?;
        if response_error(&resp).is_some() {
            return Err(AcpError(format!("initialize failed: {}", resp)));
        }

        // session/new
        let cwd_str = cwd.to_str()
            .ok_or_else(|| AcpError("cwd not utf-8".into()))?;
        let id = s.send("session/new", build_session_new_params(cwd_str))?;
        let resp = s.read_until_id(id)?;
        s.session_id = extract_session_id(&resp)
            .ok_or_else(|| AcpError(format!("no sessionId: {}", resp)))?
            .to_string();

        Ok(s)
    }

    /// Send a text prompt. Returns the full response text.
    pub fn prompt(&mut self, text: &str) -> Result<String, AcpError> {
        let params = build_prompt_params(&self.session_id, text);
        let id = self.send("session/prompt", params)?;
        self.collect_response(id)
    }

    /// Send a text prompt with @file references prepended.
    pub fn prompt_with_files(&mut self, text: &str, files: &[&str]) -> Result<String, AcpError> {
        self.prompt(&build_prompt_text(text, Some(files)))
    }

    fn send(&mut self, method: &str, params: Value) -> Result<u64, AcpError> {
        let id = self.next_id;
        self.next_id += 1;
        let line = build_request(id, method, params);
        self.stdin.write_all(line.as_bytes())
            .and_then(|_| self.stdin.write_all(b"\n"))
            .and_then(|_| self.stdin.flush())
            .map_err(|e| AcpError(format!("write stdin: {}", e)))?;
        Ok(id)
    }

    fn read_until_id(&mut self, expected_id: u64) -> Result<Value, AcpError> {
        loop {
            let line = self.read_line()?;
            if is_response(&line, expected_id) {
                return Ok(line);
            }
        }
    }

    fn collect_response(&mut self, expected_id: u64) -> Result<String, AcpError> {
        let mut chunks = Vec::new();
        loop {
            let line = self.read_line()?;
            if is_response(&line, expected_id) {
                if let Some(err) = response_error(&line) {
                    return Err(AcpError(format!("prompt error: {}", err)));
                }
                return Ok(chunks.join("").trim().to_string());
            }
            if let Some(text) = extract_chunk_text(&line) {
                chunks.push(text.to_string());
            }
        }
    }

    fn read_line(&mut self) -> Result<Value, AcpError> {
        let mut buf = String::new();
        loop {
            buf.clear();
            let n = self.reader.read_line(&mut buf)
                .map_err(|e| AcpError(format!("read stdout: {}", e)))?;
            if n == 0 {
                return Err(AcpError("process closed stdout".into()));
            }
            let trimmed = buf.trim();
            if trimmed.is_empty() {
                continue;
            }
            return serde_json::from_str(trimmed)
                .map_err(|e| AcpError(format!("parse JSON: {} in {:?}", e, &trimmed[..trimmed.len().min(100)])));
        }
    }

    pub fn session_id(&self) -> &str {
        &self.session_id
    }
}

impl Drop for AcpSession {
    fn drop(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    // === build_request ===

    #[test]
    fn build_request_initialize() {
        let msg = build_request(1, "initialize", build_initialize_params());
        let parsed: Value = serde_json::from_str(&msg).unwrap();
        assert_eq!(parsed["jsonrpc"], "2.0");
        assert_eq!(parsed["id"], 1);
        assert_eq!(parsed["method"], "initialize");
        assert_eq!(parsed["params"]["protocolVersion"], 1);
        assert_eq!(parsed["params"]["clientInfo"]["name"], "resident-ai");
    }

    #[test]
    fn build_request_session_new() {
        let msg = build_request(2, "session/new", build_session_new_params("/tmp/work"));
        let parsed: Value = serde_json::from_str(&msg).unwrap();
        assert_eq!(parsed["method"], "session/new");
        assert_eq!(parsed["params"]["cwd"], "/tmp/work");
        assert_eq!(parsed["params"]["mcpServers"], json!([]));
    }

    #[test]
    fn build_request_prompt() {
        let msg = build_request(3, "session/prompt", build_prompt_params("uuid-123", "hello"));
        let parsed: Value = serde_json::from_str(&msg).unwrap();
        assert_eq!(parsed["method"], "session/prompt");
        assert_eq!(parsed["params"]["sessionId"], "uuid-123");
        assert_eq!(parsed["params"]["prompt"][0]["type"], "text");
        assert_eq!(parsed["params"]["prompt"][0]["text"], "hello");
    }

    // === build_prompt_text ===

    #[test]
    fn prompt_text_no_files() {
        assert_eq!(build_prompt_text("hello", None), "hello");
    }

    #[test]
    fn prompt_text_empty_files() {
        assert_eq!(build_prompt_text("hello", Some(&[])), "hello");
    }

    #[test]
    fn prompt_text_with_files() {
        assert_eq!(
            build_prompt_text("analyze", Some(&["a.jpg", "b.pdf"])),
            "@a.jpg @b.pdf analyze"
        );
    }

    // === extract_session_id ===

    #[test]
    fn extract_session_id_valid() {
        let resp = json!({"result": {"sessionId": "abc-123"}});
        assert_eq!(extract_session_id(&resp), Some("abc-123"));
    }

    #[test]
    fn extract_session_id_missing() {
        let resp = json!({"result": {}});
        assert_eq!(extract_session_id(&resp), None);
    }

    #[test]
    fn extract_session_id_error_response() {
        let resp = json!({"error": {"code": -32603, "message": "fail"}});
        assert_eq!(extract_session_id(&resp), None);
    }

    // === extract_chunk_text ===

    #[test]
    fn extract_chunk_text_valid() {
        let line = json!({
            "jsonrpc": "2.0",
            "method": "session/update",
            "params": {
                "sessionId": "uuid",
                "update": {
                    "sessionUpdate": "agent_message_chunk",
                    "content": {"type": "text", "text": "4"}
                }
            }
        });
        assert_eq!(extract_chunk_text(&line), Some("4"));
    }

    #[test]
    fn extract_chunk_text_not_a_chunk() {
        let line = json!({
            "jsonrpc": "2.0",
            "method": "session/update",
            "params": {
                "update": {
                    "sessionUpdate": "available_commands_update",
                    "availableCommands": []
                }
            }
        });
        assert_eq!(extract_chunk_text(&line), None);
    }

    #[test]
    fn extract_chunk_text_not_a_notification() {
        let line = json!({"jsonrpc": "2.0", "id": 3, "result": {"stopReason": "end_turn"}});
        assert_eq!(extract_chunk_text(&line), None);
    }

    #[test]
    fn extract_chunk_thought_is_not_message() {
        let line = json!({
            "jsonrpc": "2.0",
            "method": "session/update",
            "params": {
                "update": {
                    "sessionUpdate": "agent_thought_chunk",
                    "content": {"type": "text", "text": "thinking..."}
                }
            }
        });
        assert_eq!(extract_chunk_text(&line), None);
    }

    // === is_response / response_error ===

    #[test]
    fn is_response_matching() {
        let line = json!({"jsonrpc": "2.0", "id": 5, "result": {}});
        assert_eq!(is_response(&line, 5), true);
    }

    #[test]
    fn is_response_wrong_id() {
        let line = json!({"jsonrpc": "2.0", "id": 5, "result": {}});
        assert_eq!(is_response(&line, 3), false);
    }

    #[test]
    fn is_response_notification() {
        let line = json!({"jsonrpc": "2.0", "method": "session/update", "params": {}});
        assert_eq!(is_response(&line, 1), false);
    }

    #[test]
    fn response_error_present() {
        let line = json!({"jsonrpc": "2.0", "id": 1, "error": {"code": -32601, "message": "not found"}});
        assert!(response_error(&line).is_some());
    }

    #[test]
    fn response_error_absent() {
        let line = json!({"jsonrpc": "2.0", "id": 1, "result": {}});
        assert!(response_error(&line).is_none());
    }

    // === Live tests (require gemini CLI) ===

    #[test]
    #[ignore]
    fn handshake_and_prompt() {
        let cwd = std::env::current_dir().unwrap();
        let mut session = AcpSession::new(&cwd).expect("handshake failed");
        assert!(!session.session_id().is_empty());

        let result = session.prompt("2+2は？数字だけ答えろ").expect("prompt failed");
        assert!(result.contains('4'), "expected 4, got: {}", result);
    }

    #[test]
    #[ignore]
    fn sequential_prompts() {
        let cwd = std::env::current_dir().unwrap();
        let mut session = AcpSession::new(&cwd).expect("handshake failed");

        let r1 = session.prompt("2+3は？数字だけ").expect("prompt 1 failed");
        assert!(r1.contains('5'), "expected 5, got: {}", r1);

        let r2 = session.prompt("7*8は？数字だけ").expect("prompt 2 failed");
        assert!(r2.contains("56"), "expected 56, got: {}", r2);
    }
}

File: pkg/acp/acp.go
package acp

import (
	"bufio"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/mattn/go-isatty"
	"golang.org/x/sys/windows"
	"golang.org/x/text/encoding/japanese"
	"golang.org/x/text/transform"
)

// ErrWriter returns a writer that automatically converts UTF-8 to CP932 on Windows terminals
func ErrWriter() io.Writer {
	if runtime.GOOS != "windows" {
		return os.Stderr
	}

	// Detect console output code page.
	// 932 is the standard code page for Japanese Windows (Shift-S/CP932).
	cp, _ := windows.GetConsoleOutputCP()
	if cp == 932 {
		return transform.NewWriter(os.Stderr, japanese.ShiftJIS.NewEncoder())
	}

	if isatty.IsTerminal(os.Stderr.Fd()) || isatty.IsCygwinTerminal(os.Stderr.Fd()) {
		return transform.NewWriter(os.Stderr, japanese.ShiftJIS.NewEncoder())
	}
	return os.Stderr
}

// AcpError represents a JSON-RPC error from the agent.
type AcpError struct {
	Message string
}

func (e *AcpError) Error() string {
	return e.Message
}

// Request represents a JSON-RPC 2.0 request.
type Request struct {
	JsonRPC string      `json:"jsonrpc"`
	ID      uint64      `json:"id"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params"`
}

// Response represents a JSON-RPC 2.0 response.
type Response struct {
	JsonRPC string          `json:"jsonrpc"`
	ID      uint64          `json:"id,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *ErrorDetail    `json:"error,omitempty"`
}

// ErrorDetail represents JSON-RPC error details.
type ErrorDetail struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// Notification represents a JSON-RPC 2.0 notification.
type Notification struct {
	JsonRPC string          `json:"jsonrpc"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params"`
}

// BuildRequest builds a JSON-RPC 2.0 request message.
func BuildRequest(id uint64, method string, params interface{}) string {
	req := Request{
		JsonRPC: "2.0",
		ID:      id,
		Method:  method,
		Params:  params,
	}
	data, _ := json.Marshal(req)
	return string(data)
}

// BuildInitializeParams builds the initialize request params.
func BuildInitializeParams() interface{} {
	return map[string]interface{}{
		"protocolVersion": 1,
		"clientInfo": map[string]string{
			"name":    "resident-ai-go",
			"version": "0.1.0",
		},
	}
}

// BuildSessionNewParams builds the session/new request params.
func BuildSessionNewParams(cwd string) interface{} {
	return map[string]interface{}{
		"cwd":        cwd,
		"mcpServers": []string{},
	}
}

// BuildPromptParams builds the session/prompt request params.
func BuildPromptParams(sessionID string, text string) interface{} {
	return map[string]interface{}{
		"sessionId": sessionID,
		"prompt": []map[string]string{
			{"type": "text", "text": text},
		},
	}
}

// BuildPromptText builds prompt text with @file references prepended.
func BuildPromptText(text string, files []string) string {
	if len(files) == 0 {
		return text
	}
	var sb strings.Builder
	for _, f := range files {
		sb.WriteString("@")
		sb.WriteString(f)
		sb.WriteString(" ")
	}
	sb.WriteString(text)
	return sb.String()
}

// Session represents an active ACP session with a running gemini process.
type Session struct {
	cmd       *exec.Cmd
	stdin     io.WriteCloser
	stdout    io.ReadCloser
	reader    *bufio.Reader
	sessionID string
	nextID    uint64
	model     string
}

// NewSession spawns gemini.cmd --acp, performs handshake, returns ready session.
func NewSession(cwd string) (*Session, error) {
	return NewSessionWithModel(cwd, "")
}

// NewSessionWithModel spawns gemini.cmd --acp with a specific model.
func NewSessionWithModel(cwd string, model string) (*Session, error) {
	args := []string{"--acp"}
	if model != "" {
		args = append(args, "-m", model)
	}

	cmd := exec.Command("gemini.cmd", args...)
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("stdin pipe: %w", err)
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("stdout pipe: %w", err)
	}
	cmd.Stderr = os.Stderr // Optional: redirect stderr for debugging

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("spawn gemini.cmd --acp: %w", err)
	}

	s := &Session{
		cmd:    cmd,
		stdin:  stdin,
		stdout: stdout,
		reader: bufio.NewReader(stdout),
		nextID: 1,
		model:  model,
	}

	// initialize
	id, err := s.send("initialize", BuildInitializeParams())
	if err != nil {
		s.Close()
		return nil, err
	}
	resp, err := s.readUntilID(id)
	if err != nil {
		s.Close()
		return nil, err
	}
	if resp.Error != nil {
		s.Close()
		return nil, &AcpError{Message: fmt.Sprintf("initialize failed: %s", resp.Error.Message)}
	}

	// session/new
	id, err = s.send("session/new", BuildSessionNewParams(cwd))
	if err != nil {
		s.Close()
		return nil, err
	}
	resp, err = s.readUntilID(id)
	if err != nil {
		s.Close()
		return nil, err
	}
	if resp.Error != nil {
		s.Close()
		return nil, &AcpError{Message: fmt.Sprintf("session/new failed: %s", resp.Error.Message)}
	}

	var result struct {
		SessionID string `json:"sessionId"`
	}
	if err := json.Unmarshal(resp.Result, &result); err != nil {
		s.Close()
		return nil, fmt.Errorf("parse sessionId: %w", err)
	}
	s.sessionID = result.SessionID

	return s, nil
}

// Prompt sends a text prompt. Returns the full response text.
func (s *Session) Prompt(text string) (string, error) {
	params := BuildPromptParams(s.sessionID, text)
	id, err := s.send("session/prompt", params)
	if err != nil {
		return "", err
	}
	return s.collectResponse(id)
}

// PromptWithFiles sends a text prompt with @file references prepended.
func (s *Session) PromptWithFiles(text string, files []string) (string, error) {
	return s.Prompt(BuildPromptText(text, files))
}

// PromptWithImage sends a text prompt with an image embedded as base64.
func (s *Session) PromptWithImage(text string, imagePath string) (string, error) {
	return s.PromptWithImagesInline(text, []string{imagePath})
}

// PromptWithImagesInline sends a text prompt with N images embedded as base64.
func (s *Session) PromptWithImagesInline(text string, imagePaths []string) (string, error) {
	var promptItems []interface{}
	for _, path := range imagePaths {
		data, err := os.ReadFile(path)
		if err != nil {
			return "", fmt.Errorf("read image %s: %w", path, err)
		}
		b64 := base64.StdEncoding.EncodeToString(data)
		mime := getMimeType(path)
		promptItems = append(promptItems, map[string]string{
			"type":     "image",
			"mimeType": mime,
			"data":     b64,
		})
	}
	promptItems = append(promptItems, map[string]string{
		"type": "text",
		"text": text,
	})

	params := map[string]interface{}{
		"sessionId": s.sessionID,
		"prompt":    promptItems,
	}
	id, err := s.send("session/prompt", params)
	if err != nil {
		return "", err
	}
	return s.collectResponse(id)
}

func (s *Session) send(method string, params interface{}) (uint64, error) {
	id := s.nextID
	s.nextID++
	line := BuildRequest(id, method, params)
	
	// Truncate base64 for cleaner log
	logLine := line
	if len(logLine) > 1000 {
		logLine = logLine[:1000] + "...(truncated)"
	}
	fmt.Fprintf(ErrWriter(), "ACP SEND: %s\n", logLine)
	
	_, err := s.stdin.Write([]byte(line + "\n"))
	if err != nil {
		return 0, fmt.Errorf("write stdin: %w", err)
	}
	return id, nil
}

func (s *Session) readUntilID(expectedID uint64) (*Response, error) {
	for {
		line, err := s.readLine()
		if err != nil {
			return nil, err
		}
		var resp Response
		if err := json.Unmarshal(line, &resp); err == nil && resp.ID == expectedID {
			return &resp, nil
		}
	}
}

func (s *Session) collectResponse(expectedID uint64) (string, error) {
	var chunks []string
	for {
		line, err := s.readLine()
		if err != nil {
			return "", err
		}

		// Check if it's a response
		var resp Response
		if err := json.Unmarshal(line, &resp); err == nil && resp.ID == expectedID {
			if resp.Error != nil {
				return "", &AcpError{Message: fmt.Sprintf("prompt error: %s", resp.Error.Message)}
			}
			return strings.TrimSpace(strings.Join(chunks, "")), nil
		}

		// Check if it's a notification (chunk)
		var notif Notification
		if err := json.Unmarshal(line, &notif); err == nil && notif.Method == "session/update" {
			var params struct {
				Update struct {
					SessionUpdate string `json:"sessionUpdate"`
					Content      struct {
						Text string `json:"text"`
					} `json:"content"`
				} `json:"update"`
			}
			if err := json.Unmarshal(notif.Params, &params); err == nil {
				if params.Update.SessionUpdate == "agent_message_chunk" {
					chunks = append(chunks, params.Update.Content.Text)
				}
			}
		}
	}
}

func (s *Session) readLine() ([]byte, error) {
	for {
		line, err := s.reader.ReadBytes('\n')
		if err != nil {
			return nil, fmt.Errorf("read stdout: %w", err)
		}
		trimmed := strings.TrimSpace(string(line))
		if trimmed == "" {
			continue
		}
		return []byte(trimmed), nil
	}
}

func (s *Session) Close() {
	if s.stdin != nil {
		s.stdin.Close()
	}
	if s.cmd != nil && s.cmd.Process != nil {
		s.cmd.Process.Kill()
		s.cmd.Wait()
	}
}

func (s *Session) SessionID() string {
	return s.sessionID
}

func getMimeType(path string) string {
	ext := strings.ToLower(filepath.Ext(path))
	switch ext {
	case ".png":
		return "image/png"
	case ".jpg", ".jpeg":
		return "image/jpeg"
	case ".gif":
		return "image/gif"
	case ".webp":
		return "image/webp"
	default:
		return "application/octet-stream"
	}
}
